blog/_posts/2021-12-14-sustainable-scala.md

```diff
@@ -0,0 +1,278 @@
+  also run Scala benchmarks.
+
+However, things were not that simple in practice. First, we discovered that the 
+existing infrastructure was not properly warming-up the JVM before running the 
+benchmarks, leading to an over-estimation of about 30% of their energy 
+consumption. Second, and more importantly, we noticed that the different 
+implementations of the same benchmarks (one per programming language) were 
+sometimes using different algorithms, making things obviously less comparable.
+
+We managed to find a solution to the first issue by implementing a [JVM-based 
+runner] for the benchmarks. Our runner implements the classic scheme of 
+running several warming-up iterations before doing measurements. It measures 
+the energy consumption by using [jRAPL].
+
+Dealing with the second issue was a bit less satisfactory. We don’t have the 
+capacity of checking and fixing the 50 benchmark implementations to use 
+comparable algorithms. Consequently, we only focused on the Java and Scala 
+benchmarks whose performance was too far behind their C counterpart. We 
+modified them, when necessary, to use the same algorithm as the C 
+implementation of the benchmark.
+
+After several days of work, we had 10 Scala implementations of the 
+benchmarks as well as an infrastructure properly warming-up the JVM before 
+measuring their execution time, memory consumption, and energy consumption. 
+
+## What about day-to-day code?
+
+Heavily optimized code may be different from idiomatic code 
+that we write day-to-day, and the results we would get from running the 
+benchmarks may not be applicable to idiomatic Scala code.
+
+For this reason, we also created "more idiomatic" versions of the Scala
+benchmarks. The usage of the double quotes is justified because 1) what is 
+qualified as idiomatic is fairly subjective, and 2) due to our limited 
+resources, we did not completely overhaul the implementations, but we 
+only changed some patterns that we believe are typically non-idiomatic.
+
+For instance, the "optimized" version of the `binary-trees` benchmark models 
+an empty tree as tree whose branches are `null`:
+
+~~~ scala
+final case class Tree(left: Tree, right: Tree) {
+  def checkSum: Int =
+    left match {
+      case null => 1
+      case tl   => 1 + tl.checkSum + right.checkSum
+    }
+}
+
+object Tree {
+  final val EmptyTree = Tree(null, null)
+}
+~~~
+
+Whereas the idiomatic one uses a class hierarchy:
+
+~~~ scala
+sealed trait Tree {
+  def checkSum: Int
+}
+
+case class NonEmptyTree(left: Tree, right: Tree) extends Tree {
+  def checkSum: Int = 1 + left.checkSum + right.checkSum
+}
+
+case object EmptyTree extends Tree {
+  def checkSum: Int = 1
+}
+~~~
+
+Does this change impact performance? Keep reading to know the answer.
+
+Another example of difference between idiomatic and optimized versions is the 
+usage of Scala collections instead of Java `Array`s, and the usage of `for` 
```

Comment:
Is the term `for loop`  correct in Scala? I know that you were thinking here about for-comprehension or iterating methods, like `foreach`, `map` etc. I'm wondering if it is technically correct.

Reply:
I think so… Let’s see if other programmers find it confusing as well…